pallets/funding/src/runtime_api.rs

```diff
@@ -0,0 +1,225 @@
+				let funding_target = project_details.fundraising_target_usd;
+				let funding_ratio = FixedU128::from_rational(funding_reached.into(), funding_target.into());
+				(project_id, project_details, funding_ratio)
+			})
+			.sorted_by(|a, b| b.2.cmp(&a.2))
+			.take(amount as usize)
+			.map(|(project_id, project_details, _funding_ratio)| {
+				let project_metadata = ProjectsMetadata::<T>::get(project_id).unwrap();
+				(project_id, project_metadata, project_details)
+			})
+			.collect_vec()
+	}
+
+	pub fn contribution_tokens(account: AccountIdOf<T>) -> Vec<(ProjectId, BalanceOf<T>)> {
+		let asset_ids = <T as Config>::ContributionTokenCurrency::asset_ids();
+		asset_ids
+			.filter_map(|asset_id| {
+				let balance = <T as Config>::ContributionTokenCurrency::balance(asset_id, &account);
+				if balance > Zero::zero() {
+					Some((asset_id, balance))
+				} else {
+					None
+				}
+			})
+			.sorted_by(|a, b| b.1.cmp(&a.1))
+			.collect_vec()
+	}
+
+	pub fn funding_asset_to_ct_amount(
+		project_id: ProjectId,
+		asset: AcceptedFundingAsset,
+		asset_amount: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		let project_details = ProjectsDetails::<T>::get(project_id).unwrap();
+		let funding_asset_id = asset.to_assethub_id();
+		let funding_asset_decimals = T::FundingCurrency::decimals(funding_asset_id);
+		let funding_asset_usd_price =
+			T::PriceProvider::get_decimals_aware_price(funding_asset_id, USD_DECIMALS, funding_asset_decimals).unwrap();
+		let usd_ticket_size = funding_asset_usd_price.saturating_mul_int(asset_amount);
+
+		let mut ct_amount = Zero::zero();
+
+		// Contribution phase
+		if let Some(wap) = project_details.weighted_average_price {
+			ct_amount = wap.reciprocal().unwrap().saturating_mul_int(usd_ticket_size);
+		}
+		// Auction phase, we need to consider multiple buckets
+		else {
+			let mut usd_to_spend = usd_ticket_size;
+			let mut current_bucket = Buckets::<T>::get(project_id).unwrap();
+			while usd_to_spend > Zero::zero() {
+				let bucket_price = current_bucket.current_price;
+				dbg!(&bucket_price);
+				dbg!(&usd_to_spend);
+
+				let ct_to_buy = bucket_price.reciprocal().unwrap().saturating_mul_int(usd_to_spend);
+				dbg!(&ct_to_buy);
+				let ct_to_buy = ct_to_buy.min(current_bucket.amount_left);
+				dbg!(&ct_to_buy);
+
+				ct_amount = ct_amount.saturating_add(ct_to_buy);
+				// if usd spent is 0, we will have an infinite loop
+				let usd_spent = bucket_price.saturating_mul_int(ct_to_buy).max(One::one());
+				dbg!(&usd_spent);
+				usd_to_spend = usd_to_spend.saturating_sub(usd_spent);
+				dbg!(&usd_to_spend);
+				dbg!(&ct_amount);
+
+				current_bucket.update(ct_to_buy)
+			}
+		}
+
+		ct_amount
+	}
+
+	pub fn all_project_participations_by_did(project_id: ProjectId, did: Did) -> Vec<ProjectParticipationIds<T>> {
+		let evaluations = Evaluations::<T>::iter_prefix((project_id,))
+			.filter(|((_account_id, _evaluation_id), evaluation)| evaluation.did == did)
```

Comment:
Not tested. but we should be able to abstract the filtering logic, with something like:

```rust
fn collect_ids_by_did<T, I>(iter: I, did: Did) -> Vec<(AccountIdOf<T>, u32)>
where
    I: Iterator<Item = ((AccountIdOf<T>, u32), Item)>,
    Item: { fn did(&self) -> &Did }
{
    iter
        .filter(|((_account_id, _id), item)| item.did() == &did)
        .map(|((account_id, id), _)| (account_id, id))
        .collect_vec()
}
```

Usage:

```rust
let evaluations = collect_ids_by_did(Evaluations::<T>::iter_prefix((project_id,)), did);
    let bids = collect_ids_by_did(Bids::<T>::iter_prefix((project_id,)), did);
```

EDIT: It should be possible to use it also in the `projects_by_did` API

Reply:
I would need to implement a method on every item type to extract the did, plus all that generic function. Overkill for a couple simple iterator methods.

Happy to revisit this after the simplification PRs